torch/csrc/autograd/variable.h

```diff
@@ -1,329 +1,615 @@
 #pragma once
 
-// A wrapper around at::Tensor to represent autograd Variables. Variables
-// can be implicitly converted to an at::Tensor.
+#include <Python.h>
 
-#include <mutex>
+#include "torch/csrc/autograd/edge.h"
+#include "torch/csrc/autograd/function_hook.h"
+#include "torch/csrc/autograd/variable_version.h"
+#include "torch/csrc/utils/auto_unique_ptr.h"
+
+#include <ATen/ATen.h>
+
+#include <list>
 #include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
 #include <vector>
-#include <functional>
-#include <ATen/ATen.h>
 
-#include "torch/csrc/assertions.h"
-#include "torch/csrc/jit/ir.h"
-#include "torch/csrc/jit/tracer_state.h"
-#include "torch/csrc/autograd/function_hook.h"
-#include "torch/csrc/utils/auto_unique_ptr.h"
-#include "torch/csrc/autograd/variable_version.h"
-#include "torch/csrc/autograd/edge.h"
-#include "torch/csrc/Types.h"
+namespace torch {
+namespace autograd {
+struct Function;
+} // namespace autograd
+namespace jit { namespace tracer {
+// Has to be forward declared because tracer_state.h has a dependency on
+// variable.h.
+struct ValueTracingStateElem;
+using ValueTracingState = std::list<ValueTracingStateElem>;
+}} // namespace jit::tracer
+} // namespace torch
 
 namespace torch { namespace autograd {
 
-using at::Tensor;
-struct VariableImpl;
+///~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+///                                Variable
+///~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// A `Variable` augments a `Tensor` with the ability to interact in our
+/// autograd machinery. Conceptually, `Variable`s travel along `Edge`s between
+/// `Function`s in the autograd graph. A `Variable` can either be a leaf, like a
+/// weight in a neural network, or an interior variable, when it is the result
+/// of an operation between variables. Every `Variable` also stores another
+/// `Variable` (recursively) called its `grad` (gradient). If the variable is a
```

Comment:
it's rarely recursive, since `.grad`s are usually not leaves, and so they don't get gradient saved by default (`.grad.grad` is not a very useful value - it's not the same as taking the second-order derivatives)

Reply:
Ok, thanks for clarifying. I'll remove the word